dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1758,8 +1759,7 @@ RtpsUdpDataLink::RtpsReader::process_heartbeat_i(const RTPS::HeartBeatSubmessage
 
   // Only valid heartbeats (see spec) will be "fully" applied to writer info
   if (!(hb_first < 1 || hb_last < 0 || hb_last < hb_first.previous())) {
-    // TODO: This check breaks interoperability.
-    if (writer->first_valid_hb_ && directed && heartbeat.readerId == id_.entityId) {
+    if (writer->first_valid_hb_ && (directed || !writer->sends_directed_hb())) {
```

Comment:
```suggestion
    if (writer->first_valid_hb_ && ((directed && heartbeat.readerId == id_.entityId) || !writer->sends_directed_hb())) {
```

Reply:
This was deliberate.  The only calls to `RtpsReader::process_heartbeat_i` result from undirected or directed-to-this-Reader submessages.  The `directed` parameter is also unnecessary but I didn't remove it in this PR.